common/argparse.h

```diff
@@ -209,506 +218,464 @@ namespace Carbon {
+  void AddMetaActionOption(const ArgInfo& info,
+                           llvm::function_ref<void(ArgBuilder&)> build);
+
+  void AddIntegerPositionalArg(
+      const ArgInfo& info, llvm::function_ref<void(IntegerArgBuilder&)> build);
+  void AddStringPositionalArg(
+      const ArgInfo& info, llvm::function_ref<void(StringArgBuilder&)> build);
+  void AddOneOfPositionalArg(const ArgInfo& info,
+                             llvm::function_ref<void(OneOfArgBuilder&)> build);
+
+  void AddSubcommand(const CommandInfo& info,
+                     llvm::function_ref<void(CommandBuilder&)> build);
+
+  // Subcommands can be hidden from the help listing of their parents with
+  // this setting. Hiding a subcommand doesn't disable its own help, it just
+  // removes it from the listing.
+  void HelpHidden(bool is_help_hidden);
+
+  // Exactly one of these three should be called to select and configure the
+  // kind of the built command.
+  void RequiresSubcommand();
+  void Do(ActionT action);
+  void Meta(ActionT meta_action);
 
  private:
-  class Parser;
-
-  friend auto operator<<(llvm::raw_ostream& output, ParseResult result)
-      -> llvm::raw_ostream& {
-    switch (result) {
-      case ParseResult::Error:
-        return output << "Error";
-      case ParseResult::MetaSuccess:
-        return output << "MetaSuccess";
-      case ParseResult::Success:
-        return output << "Success";
-    }
-  }
+  friend Parser;
 
-  friend auto operator<<(llvm::raw_ostream& output, ArgKind kind)
-      -> llvm::raw_ostream& {
-    switch (kind) {
-      case ArgKind::Flag:
-        return output << "Boolean";
-      case ArgKind::Integer:
-        return output << "Integer";
-      case ArgKind::String:
-        return output << "String";
-      case ArgKind::OneOf:
-        return output << "OneOf";
-      case ArgKind::MetaActionOnly:
-        return output << "MetaActionOnly";
-      case ArgKind::Invalid:
-        return output << "Invalid";
-    }
-    CARBON_FATAL() << "Corrupt argument kind!";
-  }
+  explicit CommandBuilder(Command& command, MetaPrinter& meta_printer);
 
-  friend auto operator<<(llvm::raw_ostream& output, CommandKind kind)
-      -> llvm::raw_ostream& {
-    switch (kind) {
-      case CommandKind::Invalid:
-        return output << "Invalid";
-      case CommandKind::RequiresSubcommand:
-        return output << "RequiresSubcommand";
-      case CommandKind::Action:
-        return output << "Action";
-      case CommandKind::MetaAction:
-        return output << "MetaAction";
-    }
-    CARBON_FATAL() << "Corrupt command kind!";
-  }
+  auto AddArgImpl(const ArgInfo& info, ArgKind kind) -> Arg&;
+  void AddPositionalArgImpl(const ArgInfo& info, ArgKind kind,
+                            llvm::function_ref<void(Arg&)> build);
+  void Finalize();
+
+  Command& command_;
+  MetaPrinter& meta_printer_;
+
+  llvm::SmallDenseSet<llvm::StringRef> arg_names;
+  llvm::SmallDenseSet<llvm::StringRef> subcommand_names;
```

Comment:
Minor style, but looks like the suffix underscore is missing for these members?

Reply:
Doh, yep. Fixed.